### PR TITLE
chore(build): migrate to pyproject

### PR DESCRIPTION
**Pull request by SauravMaheshkar**
Ref: #1085 

* Move requirement files to `pyproject.toml`
* Updated GA to cache deps from `pyproject.toml`
* Update python packaging using `build`
* drop `MANIFEST.in` in favour of `tool.setuptools` in `pyproject.toml`
* quick fix to `lightly/models/modules/ijepa.py` to silence black warnings

### CODE REVIEW

**Comment by philippmwirth** [comment]
Hey @SauravMaheshkar thanks for the contribution! I will take a look ASAP.

**Comment by philippmwirth** [comment]
It looks like we will have to fix the numpy dependency problems first. Will take a look at this tomorrow.

**Comment by philippmwirth** [comment]
Hey @SauravMaheshkar , we decided to temporarily limit the numpy version to `< 2.0` can you update the pull request accordingly please?

**Comment by SauravMaheshkar** [comment]
@philippmwirth fixed in [3f542be](https://github.com/lightly-ai/lightly/pull/1559/commits/3f542bebdc8ea554ad6c841c4f203042d568a600), CI looking green now 😄

File: setup.py [hunk 1]
```diff
@@ -1,130 +1,4 @@
...
-        setup_requires=setup_requires,
-        install_requires=install_requires,
-        extras_require=extras_require,
-        python_requires=python_requires,
-        packages=packages,
-        classifiers=classifiers,
-        include_package_data=True,
-        project_urls=project_urls,
-    )
+# for packaging tools not supporting, e.g., PEP 517, PEP 660
```
Comment (philippmwirth) [inline setup.py:3]: Let's make this comment more clear. When do we need this case and what are the consequences of keeping it? Please also mention why it's not necessary to duplicate all the information that was removed above.

File: pyproject.toml [hunk 2]
```diff
@@ -1,5 +1,166 @@
...
+
+[tool.setuptools.dynamic]
+readme = {file = ["README.md"], content-type = "text/markdown"}
+version = {attr = "lightly.__version__"}
+
+[tool.setuptools.package-data]
+lightly = ["lightly/cli/config/*.yaml"]
+
+[tool.setuptools.exclude-package-data]
+lightly = ["tests/"]
```
Comment (philippmwirth) [inline pyproject.toml:163]: We should also exclude benchmarks, docs, and examples.

File: pyproject.toml [hunk 3]
```diff
@@ -1,5 +1,166 @@
...
+  "torch",
+  "torchvision",
+  "pydantic >= 1.10.5, < 2",
+  # Note: pytorch_lightning>=1.5 is required for CLI
+  # https://github.com/lightly-ai/lightly/issues/912
+  "pytorch_lightning>=1.0.4",
+  "urllib3 >= 1.25.3",
+  "aenum >= 3.1.11"
+]
+dynamic = ["version", "readme"]
```
Comment (philippmwirth) [inline pyproject.toml:55]: This is missing description, classifiers, and entry-points.

Comment (philippmwirth) [reply pyproject.toml:55]: Source: https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata

File: pyproject.toml [hunk 4]
```diff
@@ -1,5 +1,166 @@
...
+build-backend = "setuptools.build_meta"
+
+[project]
+name="lightly"
+requires-python = ">=3.6"
+authors = [
+    {name = "Lightly Team", email = "team@lightly.ai"},
+]
+license = {file = "LICENSE.txt"}
+description="A deep learning package for self-supervised learning"
```
Comment (philippmwirth) [inline pyproject.toml:15]: The `long_description` got lost. Does `pyproject.toml` support it?

**Review by philippmwirth — CHANGES_REQUESTED** [review]
Thanks a lot this looks great! I think we will have to make sure that build and release is consistent before and after merging this. I have left a few remarks.

File: pyproject.toml (hunk 4 above)
Comment (SauravMaheshkar) [reply pyproject.toml:15]: I believe the following is all that we can do in a `pyproject.toml`:

```toml
dynamic = ["version", "readme"]
# ...
[tool.setuptools.dynamic]
readme = {file = ["README.md"], content-type = "text/markdown"}
```

File: pyproject.toml (hunk 3 above)
Comment (SauravMaheshkar) [reply pyproject.toml:55]: description and classifiers are specified in:

https://github.com/SauravMaheshkar/lightly/blob/3f542bebdc8ea554ad6c841c4f203042d568a600/pyproject.toml#L15-L35

and the `console_scripts` in the `[project.scripts]` section.

File: setup.py (hunk 1 above)
Comment (SauravMaheshkar) [reply setup.py:3]: It is specified in https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html

Under an "Important" banner

File: pyproject.toml (hunk 2 above)
Comment (SauravMaheshkar) [reply pyproject.toml]: Added in [e596568](https://github.com/lightly-ai/lightly/pull/1559/commits/e5965684001e2d72d682743f45af1c7ba992dec6)

File: setup.py (hunk 1 above)
Comment (philippmwirth) [reply setup.py:3]: Ah, yes. Let's then clarify the comment:
```suggestion
# For compatibility with legacy builds or versions of tools 
# that don’t support certain packaging standards (e.g. PEP 517
# or PEP 660). The configuration is kept in pyproject.toml
```

**Comment by SauravMaheshkar** [comment]
@philippmwirth It might also be worth adding tox configuration to `pyproject.toml` viz. https://tox.wiki/en/stable/config.html#pyproject-toml

**Review by philippmwirth — APPROVED** [review]

**Pull request merged**